homeassistant/components/modbus/cover.py

```diff
@@ -0,0 +1,238 @@
+        self._state_opening = config[CONF_STATE_OPENING]
+        self._status_register = config.get(CONF_STATUS_REGISTER)
+        self._status_register_type = config[CONF_STATUS_REGISTER_TYPE]
+        self._scan_interval = timedelta(seconds=config[CONF_SCAN_INTERVAL])
+        self._value = None
+        self._available = True
+
+        # If we read cover status from coil, and not from optional status register,
+        # we interpret boolean value False as closed cover, and value True as open cover.
+        # Intermediate states are not supported in such a setup.
+        if self._coil is not None and self._status_register is None:
+            self._state_closed = False
+            self._state_open = True
+            self._state_closing = None
+            self._state_opening = None
+
+        # If we read cover status from the main register (i.e., an optional
+        # status register is not specified), we need to make sure the register_type
+        # is set to "holding".
+        if self._register is not None and self._status_register is None:
+            self._status_register = self._register
+            self._status_register_type = CALL_TYPE_REGISTER_HOLDING
+
+    async def async_added_to_hass(self):
+        """Handle entity which will be added."""
+        state = await self.async_get_last_state()
+        if not state:
+            return
+        self._value = state.state
+
+        async_track_time_interval(
+            self.hass, lambda arg: self.update(), self._scan_interval
+        )
+
+    @property
+    def device_class(self) -> Optional[str]:
+        """Return the device class of the sensor."""
+        return self._device_class
+
+    @property
+    def name(self):
+        """Return the name of the switch."""
+        return self._name
+
+    @property
+    def supported_features(self):
+        """Flag supported features."""
+        return SUPPORT_OPEN | SUPPORT_CLOSE
+
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        return self._available
+
+    @property
+    def is_opening(self):
+        """Return if the cover is opening or not."""
+        return self._value == self._state_opening
+
+    @property
+    def is_closing(self):
+        """Return if the cover is closing or not."""
+        return self._value == self._state_closing
+
+    @property
+    def is_closed(self):
+        """Return if the cover is closed or not."""
+        return self._value == self._state_closed
+
+    @property
+    def should_poll(self):
+        """Return True if entity has to be polled for state.
+
+        False if entity pushes its state to HA.
+        """
+
+        # Handle polling directly in this entity
+        return False
+
+    def open_cover(self, **kwargs: Any) -> None:
+        """Open cover."""
+        if self._coil is not None:
+            self._write_coil(True)
+        else:
+            self._write_register(self._state_open)
```

Comment:
When `should_poll` is false, we need to schedule the state update specifically in the actuator methods if there's no feedback from the device.

Reply:
I'm not sure what exactly should I do. At the moment, I schedule timer during initialization by calling `async_track_time_interval()`. Is there anything else I should schedule? What do you mean by the *actuator* methods?

Reply:
`open_cover` is an actuator method, `close_over` etc as well. These will ask the device to do something, ie actuate.

After the device has done something we want home assistant to show the new state of the device in the entity representing that device feature. So we need to update the entity state with new info.

When default polling is on, the `update` method is called automatically by home assistant and then the entity is state is updated. When default polling is off, this doesn't happen and we need to update entity state manually as needed.

To update the entity state from sync context we can call `self.schedule_update_ha_state` from inside the entity.

Reply:
I added the call to `self._update()` into the `open_cover()` and `close_cover()` methods. These methods only initiate some action on Cover, but the result is not immediate. The `self._update()` method fetches the current state, and on last line it stores the state by calling `self.schedule_update_ha_state()`